### ISSUE TITLE
NaN gradients at the rotation singularity: quaternion_exp_to_log and euler_from_quaternion differentiate through acos/asin clamped to the domain boundary

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4007]
`quaternion_exp_to_log` returns a **NaN** gradient at the identity quaternion `(1, 0, 0, 0)`. The forward value is correct — the log of the identity is the origin — but backward is poisoned, at the single most common input in the domain.

```python
>>> q = torch.tensor([1., 0., 0., 0.], requires_grad=True); quaternion_exp_to_log(q).sum().backward(); q.grad
tensor([nan, 0., 0., 0.])
```

The source is an unguarded `acos` at the boundary of its domain:

```python
# quaternion_exp_to_log, kornia/geometry/conversions.py:1191-1193
quaternion_log: torch.Tensor = (
    quaternion_vector * torch.acos(torch.clamp(quaternion_scalar, min=-1.0, max=1.0)) / norm_q
)
```

`torch.clamp(w, -1, 1)` guards the **forward** value, so `acos` never sees an out-of-domain argument and the returned value is right. But clamp passes its gradient straight through at the endpoints, and `d/dw acos(w) = -1/sqrt(1 - w²)` is `-inf` at `w = ±1`. At the identity the vector part is exactly zero, so the `-inf` meets a `0` and `0 · inf = NaN`.

This is the same failure *class* as #3949 — a singularity where `torch.where`/clamp fixes the value but not the derivative — but a different mechanism (`acos` at `±1` rather than `sqrt(0)`) in a function that pass did not cover.

Two further observations from the repro below:

- With a non-zero vector part and `w = 1` (a non-unit input) the NaN degrades to a bare `-inf` rather than becoming finite, so the defect is not specific to the all-zero vector.
- At the antipode `(-1, 0, 0, 0)` the NaN in the scalar slot is joined by `3.1416e+08` in the vector slots. That is `acos(-1)/eps = π/1e-8`, a second, `eps`-scale artifact of dividing by the clamped `norm_q` at a point where the true norm is zero.

## The same pattern at a second site

`euler_from_quaternion` has the identical `clamp` + arc-trig construction, and fails the same way once the argument lands exactly on the boundary (gimbal lock, `pitch = ±π/2`):

```python
# euler_from_quaternion, kornia/geometry/conversions.py:1376-1378
sinp = 2.0 * (w * y - z * x)
sinp = sinp.clamp(min=-1.0, max=1.0)
pitch = sinp.asin()
```

`d/dx asin(x) = 1/sqrt(1 - x²)` is `+inf` at `x = ±1`, and the products in the chain rule turn some of those into NaN. Note this one only bites when `sinp` is *exactly* `±1`; a quaternion that is at gimbal lock only to within rounding gets a finite but enormous gradient (~`1.3e16`) instead, which is arguably worse because it is silent.

## Repro

```python
import torch
from kornia.geometry.conversions import euler_from_quaternion, quaternion_exp_to_log

# 1. quaternion_exp_to_log at the identity: acos'(1) = -inf, vector part = 0 -> 0 * inf = NaN
q = torch.tensor([1.0, 0.0, 0.0, 0.0], dtype=torch.float64, requires_grad=True)
quaternion_exp_to_log(q).sum().backward()
print("exp_to_log grad at the identity (1,0,0,0)  :", q.grad)

# the antipode is the same singularity
qa = torch.tensor([-1.0, 0.0, 0.0, 0.0], dtype=torch.float64, requires_grad=True)
quaternion_exp_to_log(qa).sum().backward()
print("exp_to_log grad at the antipode (-1,0,0,0) :", qa.grad)

# with a non-zero vector part the NaN becomes a bare -inf, still unusable
qv = torch.tensor([1.0, 0.1, 0.0, 0.0], dtype=torch.float64, requires_grad=True)
quaternion_exp_to_log(qv).sum().backward()
print("exp_to_log grad at (1, 0.1, 0, 0)          :", qv.grad)

# finite away from the singularity
qg = torch.tensor([0.9, 0.3, 0.2, 0.1], dtype=torch.float64, requires_grad=True)
quaternion_exp_to_log(qg).sum().backward()
print("exp_to_log grad at (0.9,0.3,0.2,0.1)       :", qg.grad)

# 2. the same clamp+arc-trig pattern in euler_from_quaternion, at gimbal lock
w, x, y, z = (torch.tensor(v, dtype=torch.float64, requires_grad=True) for v in (1.0, 0.0, 0.5, 0.0))
out = euler_from_quaternion(w, x, y, z)
print("euler pitch at sinp == 1.0                 :", out[1].item())
sum(o.sum() for o in out).backward()
print("euler grads (w,x,y,z) at sinp == 1.0       :", [t.grad.item() for t in (w, x, y, z)])

# 3. the derivative that drives both
for s in (1.0, -1.0, 0.999999):
    t = torch.tensor(s, dtype=torch.float64, requires_grad=True)
    torch.acos(torch.clamp(t, -1.0, 1.0)).backward()
    print(f"d/ds acos(clamp(s,-1,1)) at s = {s:<9}  : {t.grad.item()}")
```

## Observed

```
exp_to_log grad at the identity (1,0,0,0)  : tensor([nan, 0., 0., 0.], dtype=torch.float64)
exp_to_log grad at the antipode (-1,0,0,0) : tensor([       nan, 3.1416e+08, 3.1416e+08, 3.1416e+08], dtype=torch.float64)
exp_to_log grad at (1, 0.1, 0, 0)          : tensor([-inf, 0., 0., 0.], dtype=torch.float64)
exp_to_log grad at (0.9,0.3,0.2,0.1)       : tensor([-3.6788, -0.3444,  0.1722,  0.6888], dtype=torch.float64)
euler pitch at sinp == 1.0                 : 1.5707963267948966
euler grads (w,x,y,z) at sinp == 1.0       : [inf, nan, inf, nan]
d/ds acos(clamp(s,-1,1)) at s = 1.0        : -inf
d/ds acos(clamp(s,-1,1)) at s = -1.0       : -inf
d/ds acos(clamp(s,-1,1)) at s = 0.999999   : -707.1069579492319
```

## Expected

A finite gradient at the identity and at the antipode, and at gimbal lock for `euler_from_quaternion`.

On the unit sphere — the intended domain — there is no mathematical singularity in the quantity being differentiated, only in the parametrization. `acos(w)` has an infinite derivative at `w = ±1`, but it is multiplied by a vector part that goes to zero there: as `θ → 0`, `acos(w) → θ/2` and `‖v‖ → θ/2`, so the log map approaches the identity map and its derivative is bounded. Sampling unit quaternions at `θ = 1e-1 … 1e-6` returns exactly `θ/2`, linearly, with no sign of a blow-up in the value. The composite is smooth; only the chosen decomposition is not.

(The `-inf` at the non-unit `(1, 0.1, 0, 0)` above is a related but distinct symptom — off the unit sphere `w = 1` with `v ≠ 0` is not a rotation at all, so there is no limit argument to appeal to there. It is included only to show that the boundary derivative, not the zero vector part, is what drives the failure.)

The fix pattern is already in this file, cited in #3949: `axis_angle_to_rotation_matrix` writes `torch.sqrt(theta2.clamp(min=1e-12))  # clamping to ensure no nan gradients`. The analogue here is to keep the `acos`/`asin` argument strictly inside `(-1, 1)` — clamping to `±(1 - eps)` rather than `±1` — or to select the singular branch with a `torch.where` whose *both* arms are finite, which is the approach #3980 took for #3949.

Why this matters: `quaternion_exp_to_log` is the log map used to put a rotation into a minimal 3-parameter tangent space, which is precisely what pose-optimization and rotation-regression loops differentiate through. The identity is the standard initialization for those loops, so the first backward pass returns NaN and the optimization dies before it starts — silent in forward, total in backward, exactly as in #3949.

## Environment

```
kornia 0.9.0rc1 (main @ 0685b1f5) | torch 2.9.1 | python 3.11.14 | CPU
```

Reproduces at float32 and float64.

---
*Posted on behalf of @ducha-aiki by Claude (Opus 5).*
